src/main/java/io/camunda/testing/assertions/ProcessInstanceAssertions.java

```diff
@@ -0,0 +1,413 @@
+            .stream()
+            .collect(Collectors.groupingBy(record -> record.getValue().getFlowScopeKey()));
+
+    if (recordsMap.isEmpty()) {
+      return false;
+    }
+
+    for (Entry<Long, List<Record<ProcessInstanceRecordValue>>> entry : recordsMap.entrySet()) {
+      final Record<ProcessInstanceRecordValue> lastRecord =
+          entry.getValue().get(entry.getValue().size() - 1);
+      if (lastRecord.getIntent() == ProcessInstanceIntent.ELEMENT_ACTIVATED) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Verifies the expectation that the process instance is currently waiting at the specified
+   * elements, and not at any other element.
+   *
+   * @param elementIds The ids of the elements
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions isWaitingExactlyAtElements(final String... elementIds) {
+    final List<String> shouldBeWaitingAt = Arrays.asList(elementIds);
+    final List<String> wrongfullyWaitingElementIds = new ArrayList<>();
+    final List<String> wrongfullyNotWaitingElementIds = new ArrayList<>();
+
+    StreamFilter.processInstance(recordStreamSource.processInstanceRecords())
+        .withProcessInstanceKey(actual.getProcessInstanceKey())
+        .withoutBpmnElementType(BpmnElementType.PROCESS)
+        .withIntent(ProcessInstanceIntent.ELEMENT_ACTIVATING).stream()
+        .map(Record::getValue)
+        .map(ProcessInstanceRecordValue::getElementId)
+        .forEach(
+            id -> {
+              if (shouldBeWaitingAt.contains(id) && !isWaitingAtElement(id)) {
+                // not waiting for the element we want to wait for!
+                wrongfullyNotWaitingElementIds.add(id);
+              } else if (!shouldBeWaitingAt.contains(id) && isWaitingAtElement(id)) {
+                // waiting for an element we don't want to wait for!
+                wrongfullyWaitingElementIds.add(id);
+              }
+            });
+
+    assertThat(wrongfullyWaitingElementIds.isEmpty())
+        .withFailMessage(
+            "Process with key %s is waiting at element(s) with id(s) %s",
+            actual.getProcessInstanceKey(), String.join(", ", wrongfullyWaitingElementIds))
+        .isTrue();
+    assertThat(wrongfullyNotWaitingElementIds.isEmpty())
+        .withFailMessage(
+            "Process with key %s is not waiting at element(s) with id(s) %s",
+            actual.getProcessInstanceKey(), String.join(", ", wrongfullyNotWaitingElementIds))
+        .isTrue();
+
+    return this;
+  }
+
+  /**
+   * Verifies the expectation that the process instance is currently waiting to receive one or more
+   * specified messages.
+   *
+   * @param messageNames Names of the messages
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions isWaitingForMessage(final String... messageNames) {
+    final List<String> notCreatedProcessMessageSubscriptions =
```

Comment:
:thought_balloon: I find this version of filter logic hard to understand/read.

Maybe it would be easier to understand like this:
```
var messageNamesOfOpenSubscriptions=  getOpenProcessMessageSubscritions().stream()
        .map(record -> record.getValue().getMessageName())
       .collect(toList();;
assertThat(messageNamesOfOpenSubscriptions).containsAll(messageNames);
```

I think the more general method `getOpenProcessMessageSubscriptions()` is the more versatile method and easier to read when you add additional logic to its result. Same for other methods e.g. waiting for elements.